Cell 1：检查 GPU

In [1]:
import torch
print("GPU 是否可用:", torch.cuda.is_available())
if torch.cuda.is_available():
    print("GPU 型号:", torch.cuda.get_device_name(0))

GPU 是否可用: True
GPU 型号: Tesla T4


Cell 2：确认数据路径

In [2]:
import os

DATA_ROOT = "/kaggle/input/datasets/wuvdji/smgc-data"
print("DATA_ROOT =", DATA_ROOT)
print("是否存在:", os.path.exists(DATA_ROOT))

for dirname, _, filenames in os.walk(DATA_ROOT):
    for filename in filenames:
        print(os.path.join(dirname, filename))


DATA_ROOT = /kaggle/input/datasets/wuvdji/smgc-data
是否存在: True
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/D1/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/D1/adata_ADT.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/A1/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Human_Lymph_Nodes/A1/adata_ADT.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/simulation/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/simulation/adata_ADT.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/E18.5_mouse_brain/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/E18.5_mouse_brain/adata_ATAC.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E18/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E18/adata_ATAC.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E15/adata_RNA.h5ad
/kaggle/input/datasets/wuvdji/smgc-data/Mouse_Embryos_S2/E15/adata_ATAC.h5ad


Cell 3：拉代码

In [3]:
# 第一步：拉代码到 /kaggle/working/SpaMGCL
!git clone https://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@github.com/huqian122/SpaMGCL.git /kaggle/working/SpaMGCL

Cloning into '/kaggle/working/SpaMGCL'...
remote: Enumerating objects: 372, done.
remote: Counting objects: 100% (372/372), done.
remote: Compressing objects: 100% (357/357), done.
remote: Total 372 (delta 162), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (372/372), 2.36 MiB | 15.60 MiB/s, done.
Resolving deltas: 100% (162/162), done.


Cell 4：安装依赖

In [4]:
%cd /kaggle/working/SpaMGCL/SpaMGCL
!pip install -q anndata scanpy

/kaggle/working/SpaMGCL/SpaMGCL
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 188.8/188.8 kB 13.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 69.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.1/79.1 kB 5.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.0/40.0 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 93.6 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.7/16.7 MB 88.6 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.6/45.6 kB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.7/363.7 kB 21.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.9/59.9 MB 32.5 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.2/9.2 MB 107.5 MB/s eta 0:00:0000:0100:01
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the f

Cell 5：冒烟测试

In [6]:
%cd /kaggle/working/SpaMGCL/SpaMGCL
!python scripts/phase4_smoke.py

/kaggle/working/SpaMGCL/SpaMGCL
L_rec=0.250091
L_MGCL=3.087323
L_cluster=2.233750
L_spatial=0.000285
WD=[0.33151009678840637, 0.17174333333969116, 0.15729603171348572, 0.13252972066402435]
SC=[0.00014662364264950156, 2.6803680157172494e-05, 4.2885065340669826e-05, 6.826334720244631e-05]
w_WD=[0.2181641012430191, 0.25595822930336, 0.2596829831600189, 0.2661946713924408]
w_SC=[0.21815554797649384, 0.255963534116745, 0.25968626141548157, 0.2661946713924408]
✅ Phase 4 验证通过


Cell 6：静态检查第一批改动

In [14]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

!python -m py_compile \
  src/losses/sample_contrastive.py \
  src/models/spamgcl.py \
  experiments/run_exp.py

print("✅ 静态检查通过")

/kaggle/working/SpaMGCL/SpaMGCL
✅ 静态检查通过


Cell 7：确认新字段已经进代码

In [15]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import subprocess

# 确认关键字段真的出现在文件里
checks = {
    "mgcl_weight_std": "src/models/spamgcl.py",
    "neg_count":       "src/losses/sample_contrastive.py",
    "sc_enabled":      "experiments/run_exp.py",
    "snf_enabled":     "experiments/run_exp.py",
    "lambda_spatial":  "experiments/run_exp.py",
    "spatial_loss":    "experiments/run_exp.py",
}

for key, path in checks.items():
    out = subprocess.run(["grep", "-n", key, path],
                         capture_output=True, text=True).stdout.strip()
    print(f"[{key}] in {path}:")
    print(out if out else "  ❌ 未找到")
    print()

/kaggle/working/SpaMGCL/SpaMGCL
[mgcl_weight_std] in src/models/spamgcl.py:
32:    mgcl_weight_std: Tensor
215:            mgcl_weight_std=weights.detach().std(unbiased=False),

[neg_count] in src/losses/sample_contrastive.py:
112:        debug["neg_count"] = int(valid_negative_mask.sum().item())

[sc_enabled] in experiments/run_exp.py:
554:            "sc_enabled": spatial_weighting_enabled,
579:            f"sc_enabled={record['sc_enabled']} | "

[snf_enabled] in experiments/run_exp.py:
555:            "snf_enabled": spatial_negative_filter_enabled,
580:            f"snf_enabled={record['snf_enabled']} | "

[lambda_spatial] in experiments/run_exp.py:
300:    lambda_spatial = float(
301:        loss.get("lambda_spatial", config.get("lambda_spatial", 0.0))
303:    if min(lambda_rec, lambda_mgcl, lambda_cluster, lambda_spatial) < 0:
322:        lambda_spatial if spatial_enabled else 0.0,
489:            lambda_spatial,
509:            lambda_spatial=lambda_spatial,
549:            "lamb

Cell 8：生成 HLN-A1 三组诊断配置

In [16]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import os, yaml
from copy import deepcopy

DATA_ROOT = "/kaggle/input/datasets/wuvdji/smgc-data"
BASE_CFG = "configs/hlna1_best.yaml"

with open(BASE_CFG) as f:
    base = yaml.safe_load(f)

os.makedirs("configs/generated", exist_ok=True)

def make(name, sc, snf, lam_spatial, start_epoch, epochs=50):
    cfg = deepcopy(base)

    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = DATA_ROOT
    cfg["data"]["label_key"] = "Spatial_Label"
    cfg["output"] = {"root": "results"}
    cfg.pop("dir", None)

    cfg["training"]["epochs"] = epochs
    cfg["training"]["spatial_start_epoch"] = start_epoch

    cfg["model"]["fine_dim"] = 32
    cfg["model"]["cluster_temperature"] = 1.0

    cfg["loss"]["lambda_cluster"] = 0.1
    cfg["loss"]["lambda_mgcl"] = 1.0
    cfg["loss"]["lambda_spatial"] = lam_spatial

    cfg["spatial"] = {
        "enabled": bool(sc or snf),
        "consistency_weighting": bool(sc),
        "negative_filter": bool(snf),
    }

    path = f"configs/generated/{name}.yaml"
    with open(path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False, allow_unicode=True)
    print(f"✅ {path}  SC={sc} SNF={snf} λ_sp={lam_spatial} start={start_epoch}")

# core：完全关闭空间机制
make("diag_hlna1_core", sc=False, snf=False, lam_spatial=0.0, start_epoch=999)

# sp0：spatial 从第 1 轮开始记录，但 λ_spatial=0（对照组，验证 spatial 有值但不进 total）
make("diag_hlna1_sp0",  sc=False, snf=False, lam_spatial=0.0, start_epoch=1)

# sc：SC 开，λ_spatial=0.1
make("diag_hlna1_sc",   sc=True,  snf=False, lam_spatial=0.1, start_epoch=1)

/kaggle/working/SpaMGCL/SpaMGCL
✅ configs/generated/diag_hlna1_core.yaml  SC=False SNF=False λ_sp=0.0 start=999
✅ configs/generated/diag_hlna1_sp0.yaml  SC=False SNF=False λ_sp=0.0 start=1
✅ configs/generated/diag_hlna1_sc.yaml  SC=True SNF=False λ_sp=0.1 start=1


Cell 9：跑三组诊断（50 轮）

In [17]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import os, subprocess, json

NAMES = ["diag_hlna1_core", "diag_hlna1_sp0", "diag_hlna1_sc"]

for name in NAMES:
    cfg_path = f"configs/generated/{name}.yaml"
    res_path = f"results/{name}/metrics.json"

    if os.path.exists(res_path):
        print(f"⏭️ 已存在，跳过 {name}")
        continue

    print("=" * 70)
    print(f"🚀 开始 {name}")
    print("=" * 70)

    ret = subprocess.run(
        ["python", "experiments/run_exp.py", "--config", cfg_path],
        cwd="/kaggle/working/SpaMGCL/SpaMGCL",
        check=False,
    )
    print("返回码:", ret.returncode)

    if os.path.exists(res_path):
        with open(res_path) as f:
            d = json.load(f)
        print(f"{name}: ARI={d.get('ARI'):.4f}  NMI={d.get('NMI'):.4f}")
    else:
        print(f"⚠️ {name} 未生成 metrics.json")

/kaggle/working/SpaMGCL/SpaMGCL
🚀 开始 diag_hlna1_core


/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.887e-03 | neg_count=12134772 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400130 | rec=0.242314 | mgcl=8.157817 | cluster=2.950293 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=2.620e-03 | neg_count=12134772 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232629 | mgcl=8.156867 | cluster=2.949508 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.570e-03 | neg_count=12134772 | effC=9.962 | gradC=0.000e+00
epoch 004/050 | total=8.379530 | rec=0.223315 | mgcl=8.156215 | cluster=2.948801 | 

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.887e-03 | neg_count=12134772 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400130 | rec=0.242314 | mgcl=8.157817 | cluster=2.950293 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=2.620e-03 | neg_count=12134772 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232629 | mgcl=8.156867 | cluster=2.949508 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.570e-03 | neg_count=12134772 | effC=9.962 | gradC=0.000e+00
epoch 004/050 | total=8.379532 | rec=0.223315 | mgcl=8.156217 | cluster=2.948801 | 

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.415441 | rec=0.252383 | mgcl=8.159291 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0.1 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=4.644e-03 | neg_count=12134772 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.403193 | rec=0.242390 | mgcl=8.157875 | cluster=2.950282 | spatial_loss=0.029273 | lambda_spatial=0.1 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=2.832e-03 | neg_count=12134772 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.392089 | rec=0.232770 | mgcl=8.157004 | cluster=2.949494 | spatial_loss=0.023150 | lambda_spatial=0.1 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=5.135e-03 | neg_count=12134772 | effC=9.962 | gradC=0.000e+00
epoch 004/050 | total=8.381841 | rec=0.223514 | mgcl=8.156461 | cluster=2.948784

Cell 10：对照诊断

In [18]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import json, pandas as pd

NAMES = ["diag_hlna1_core", "diag_hlna1_sp0", "diag_hlna1_sc"]
rows = []

for name in NAMES:
    p = f"results/{name}/metrics.json"
    with open(p) as f:
        d = json.load(f)

    h = d["loss_history"][-1]

    rows.append({
        "name": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "total":        round(h.get("total", -1), 4),
        "rec":          round(h.get("reconstruction", -1), 4),
        "mgcl":         round(h.get("sample_contrastive", -1), 4),
        "mgcl_w_std":   round(h.get("mgcl_weight_std", -1), 5),
        "cluster":      round(h.get("cluster_contrastive", -1), 4),
        "spatial":      round(h.get("spatial_loss", -1), 6),
        "lambda_sp":    h.get("lambda_spatial", -1),
        "SC":           h.get("sc_enabled", None),
        "SNF":          h.get("snf_enabled", None),
        "neg_count":    h.get("neg_count", -1),
    })

df = pd.DataFrame(rows)
pd.set_option("display.width", 200)
print(df.to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
           name    ARI    NMI  total    rec   mgcl  mgcl_w_std  cluster  spatial  lambda_sp    SC   SNF  neg_count
diag_hlna1_core 0.2890 0.3534 7.5940 0.0763 7.2260     0.00480   2.9171  0.00000        0.0 False False   12134772
 diag_hlna1_sp0 0.2893 0.3536 7.5940 0.0763 7.2260     0.00481   2.9171  0.00000        0.0 False False   12134772
  diag_hlna1_sc 0.1989 0.2192 8.0598 0.0575 7.6289     0.01032   2.9370  0.79716        0.1  True False   12134772


Cell 11：自动判断是否通过

In [19]:
import json, os

def load(name):
    with open(f"results/{name}/metrics.json") as f:
        return json.load(f)

core = load("diag_hlna1_core")["loss_history"][-1]
sp0  = load("diag_hlna1_sp0")["loss_history"][-1]
sc   = load("diag_hlna1_sc")["loss_history"][-1]

print("=" * 70)
print("第一批修复验收")
print("=" * 70)

# 验收 1：SC on 是否改变 mgcl 或 mgcl_weight_std
mgcl_core = core.get("sample_contrastive", 0)
mgcl_sc   = sc.get("sample_contrastive", 0)
std_core  = core.get("mgcl_weight_std", 0)
std_sc    = sc.get("mgcl_weight_std", 0)

sc_changed = (
    abs(mgcl_sc - mgcl_core) > 1e-4
    or abs(std_sc - std_core) > 1e-4
)
print(f"[1] SC on/off 是否改变 MGCL 输入：")
print(f"    mgcl_core={mgcl_core:.5f}  mgcl_sc={mgcl_sc:.5f}  Δ={mgcl_sc - mgcl_core:+.5f}")
print(f"    std_core ={std_core:.5f}  std_sc ={std_sc:.5f}  Δ={std_sc - std_core:+.5f}")
print(f"    结果：{'✅ 通过' if sc_changed else '❌ 未通过'}")
print()

# 验收 2：spatial 有值但 λ=0 时不影响 total
sp0_spatial = sp0.get("spatial_loss", 0)
sp0_lam     = sp0.get("lambda_spatial", -1)
print(f"[2] λ_spatial=0 时 spatial 仍能记录：")
print(f"    spatial_loss={sp0_spatial:.6f}  lambda_spatial={sp0_lam}")
print(f"    结果：{'✅ 通过' if sp0_spatial > 0 and sp0_lam == 0 else '❌ 未通过'}")
print()

# 验收 3：λ_spatial=0.1 时 total 里多出 spatial 项
sc_rec  = sc.get("reconstruction", 0)
sc_mgcl = sc.get("sample_contrastive", 0)
sc_clu  = sc.get("cluster_contrastive", 0)
sc_sp   = sc.get("spatial_loss", 0)
sc_total = sc.get("total", 0)
expected = sc_rec * 1.0 + sc_mgcl * 1.0 + sc_clu * 0.1 + sc_sp * 0.1
diff = abs(sc_total - expected)

print(f"[3] λ_spatial=0.1 时 total 是否包含 spatial 项：")
print(f"    total={sc_total:.6f}  expected≈{expected:.6f}  |diff|={diff:.6f}")
print(f"    结果：{'✅ 通过' if diff < 1e-3 else '❌ 未通过（total 公式没生效或 lambda 不对）'}")
print()

print("=" * 70)
if sc_changed and sp0_spatial > 0 and sp0_lam == 0 and diff < 1e-3:
    print("🎉 第一批修复全部通过，可以发第二批 Codex（修 SNF + SNF-only）")
else:
    print("⚠️ 有项目未通过，先别进第二批，把上面表格贴给 Codex 或我")
print("=" * 70)

第一批修复验收
[1] SC on/off 是否改变 MGCL 输入：
    mgcl_core=7.22597  mgcl_sc=7.62887  Δ=+0.40290
    std_core =0.00480  std_sc =0.01032  Δ=+0.00552
    结果：✅ 通过

[2] λ_spatial=0 时 spatial 仍能记录：
    spatial_loss=0.000000  lambda_spatial=0.0
    结果：❌ 未通过

[3] λ_spatial=0.1 时 total 是否包含 spatial 项：
    total=8.059798  expected≈8.059798  |diff|=0.000000
    结果：✅ 通过

⚠️ 有项目未通过，先别进第二批，把上面表格贴给 Codex 或我


改 Cell 8 生成四组配置：

In [8]:
%cd /kaggle/working/SpaMGCL/SpaMGCL
!git pull
!git log --oneline -3

import os, yaml
from copy import deepcopy

DATA_ROOT = "/kaggle/input/datasets/wuvdji/smgc-data"
BASE_CFG = "configs/hlna1_best.yaml"

with open(BASE_CFG) as f:
    base = yaml.safe_load(f)

os.makedirs("configs/generated", exist_ok=True)

def make(name, sc, snf, lam_spatial, start_epoch, epochs=50):
    cfg = deepcopy(base)
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = DATA_ROOT
    cfg["data"]["label_key"] = "Spatial_Label"
    cfg["output"] = {"root": "results"}
    cfg.pop("dir", None)
    cfg["training"]["epochs"] = epochs
    cfg["training"]["spatial_start_epoch"] = start_epoch
    cfg["model"]["fine_dim"] = 32
    cfg["model"]["cluster_temperature"] = 1.0
    cfg["loss"]["lambda_cluster"] = 0.1
    cfg["loss"]["lambda_mgcl"] = 1.0
    cfg["loss"]["lambda_spatial"] = lam_spatial
    cfg["spatial"] = {
        "enabled": bool(sc or snf),
        "consistency_weighting": bool(sc),
        "negative_filter": bool(snf),
    }
    path = f"configs/generated/{name}.yaml"
    with open(path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False, allow_unicode=True)
    print(f"✅ {path}  SC={sc} SNF={snf} λ_sp={lam_spatial} start={start_epoch}")

make("diag_hlna1_core",   sc=False, snf=False, lam_spatial=0.0, start_epoch=999)
make("diag_hlna1_sc",     sc=True,  snf=False, lam_spatial=0.1, start_epoch=1)
make("diag_hlna1_snf",    sc=False, snf=True,  lam_spatial=0.1, start_epoch=1)
make("diag_hlna1_sc_snf", sc=True,  snf=True,  lam_spatial=0.1, start_epoch=1)

/kaggle/working/SpaMGCL/SpaMGCL
Already up to date.
1b50e80 (HEAD -> main, origin/main, origin/HEAD) Add files via upload
fbb9c7a Add files via upload
e480f6e Add files via upload
✅ configs/generated/diag_hlna1_core.yaml  SC=False SNF=False λ_sp=0.0 start=999
✅ configs/generated/diag_hlna1_sc.yaml  SC=True SNF=False λ_sp=0.1 start=1
✅ configs/generated/diag_hlna1_snf.yaml  SC=False SNF=True λ_sp=0.1 start=1
✅ configs/generated/diag_hlna1_sc_snf.yaml  SC=True SNF=True λ_sp=0.1 start=1


跑四组

In [9]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import os, subprocess, json

NAMES = ["diag_hlna1_core", "diag_hlna1_sc",
         "diag_hlna1_snf", "diag_hlna1_sc_snf"]

for name in NAMES:
    cfg_path = f"configs/generated/{name}.yaml"
    res_path = f"results/{name}/metrics.json"
    if os.path.exists(res_path):
        print("⏭️ 已存在，跳过", name); continue
    print("=" * 70, "\n🚀", name)
    ret = subprocess.run(
        ["python", "experiments/run_exp.py", "--config", cfg_path],
        cwd="/kaggle/working/SpaMGCL/SpaMGCL", check=False,
    )
    print("返回码:", ret.returncode)

/kaggle/working/SpaMGCL/SpaMGCL
🚀 diag_hlna1_core


/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.887e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400129 | rec=0.242314 | mgcl=8.157816 | cluster=2.950293 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=2.620e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232629 | mgcl=8.156867 | cluster=2.949508 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.570e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch 00

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.415441 | rec=0.252383 | mgcl=8.159291 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0.1 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=4.644e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.403193 | rec=0.242390 | mgcl=8.157875 | cluster=2.950282 | spatial_loss=0.029273 | lambda_spatial=0.1 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=2.832e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.392089 | rec=0.232770 | mgcl=8.157004 | cluster=2.949494 | spatial_loss=0.023150 | lambda_spatial=0.1 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=5.135e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.414354 | rec=0.252383 | mgcl=8.158205 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0.1 | sc_enabled=False | snf_enabled=True | mgcl_weight_std=4.887e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.402103 | rec=0.242390 | mgcl=8.156786 | cluster=2.950282 | spatial_loss=0.029275 | lambda_spatial=0.1 | sc_enabled=False | snf_enabled=True | mgcl_weight_std=2.780e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.391001 | rec=0.232770 | mgcl=8.155915 | cluster=2.949493 | spatial_loss=0.023153 | lambda_spatial=0.1 | sc_enabled=False | snf_enabled=True | mgcl_weight_std=5.010e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.962 | gradC=0.0

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.414353 | rec=0.252383 | mgcl=8.158204 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0.1 | sc_enabled=True | snf_enabled=True | mgcl_weight_std=4.644e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.402102 | rec=0.242390 | mgcl=8.156785 | cluster=2.950282 | spatial_loss=0.029273 | lambda_spatial=0.1 | sc_enabled=True | snf_enabled=True | mgcl_weight_std=2.832e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.391003 | rec=0.232770 | mgcl=8.155917 | cluster=2.949493 | spatial_loss=0.023151 | lambda_spatial=0.1 | sc_enabled=True | snf_enabled=True | mgcl_weight_std=5.135e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.962 | gradC=0.000e

对照诊断

In [10]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import json, pandas as pd

NAMES = ["diag_hlna1_core", "diag_hlna1_sc",
         "diag_hlna1_snf", "diag_hlna1_sc_snf"]
rows = []

for name in NAMES:
    d = json.load(open(f"results/{name}/metrics.json"))
    h = d["loss_history"][-1]
    rows.append({
        "name": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "total": round(h.get("total", -1), 4),
        "mgcl":  round(h.get("sample_contrastive", -1), 4),
        "mgcl_w_std": round(h.get("mgcl_weight_std", -1), 5),
        "spatial": round(h.get("spatial_loss", -1), 6),
        "λ_sp": h.get("lambda_spatial", -1),
        "SC": h.get("sc_enabled"),
        "SNF": h.get("snf_enabled"),
        "neg_count": h.get("neg_count", -1),
        "masked": h.get("snf_masked_positions", -1),
    })

df = pd.DataFrame(rows)
pd.set_option("display.width", 240)
print(df.to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
             name    ARI    NMI  total   mgcl  mgcl_w_std  spatial  λ_sp    SC   SNF  neg_count  masked
  diag_hlna1_core 0.2892 0.3536 7.5940 7.2260     0.00481 0.000000   0.0 False False   12134772       0
    diag_hlna1_sc 0.1987 0.2192 8.0598 7.6288     0.01033 0.797250   0.1  True False   12134772       0
   diag_hlna1_snf 0.1995 0.2203 8.0574 7.6265     0.00469 0.796252   0.1 False  True   12121566   13206
diag_hlna1_sc_snf 0.1988 0.2191 8.0574 7.6263     0.01030 0.797901   0.1  True  True   12121566   13206


 自动判断

In [11]:
import json

def load(n):
    return json.load(open(f"results/{n}/metrics.json"))["loss_history"][-1]

core   = load("diag_hlna1_core")
sc     = load("diag_hlna1_sc")
snf    = load("diag_hlna1_snf")
sc_snf = load("diag_hlna1_sc_snf")

print("=" * 70)
print("第二批 SNF 修复验收")
print("=" * 70)

# 验收 1：SNF off/on 时 neg_count 明显不同
print("[1] SNF off/on 是否改变 neg_count：")
print(f"    core.neg_count = {core['neg_count']}")
print(f"    snf.neg_count  = {snf['neg_count']}")
print(f"    Δ = {snf['neg_count'] - core['neg_count']}")
ok1 = snf['neg_count'] < core['neg_count']
print(f"    结果：{'✅ 通过' if ok1 else '❌ 未通过'}")
print()

# 验收 2：SNF on 时 masked_positions > 0
print("[2] SNF on 时 masked_positions > 0：")
print(f"    snf.masked    = {snf.get('snf_masked_positions', -1)}")
print(f"    sc_snf.masked = {sc_snf.get('snf_masked_positions', -1)}")
ok2 = snf.get('snf_masked_positions', 0) > 0 and sc_snf.get('snf_masked_positions', 0) > 0
print(f"    结果：{'✅ 通过' if ok2 else '❌ 未通过'}")
print()

# 验收 3：SNF-only 日志正确
print("[3] SNF-only 日志 SC=False SNF=True：")
print(f"    sc_enabled={snf['sc_enabled']}  snf_enabled={snf['snf_enabled']}")
ok3 = (snf['sc_enabled'] is False) and (snf['snf_enabled'] is True)
print(f"    结果：{'✅ 通过' if ok3 else '❌ 未通过'}")
print()

# 验收 4：total 公式
expected = sc_snf['reconstruction'] * 1.0 + sc_snf['sample_contrastive'] * 1.0 \
         + sc_snf['cluster_contrastive'] * 0.1 + sc_snf['spatial_loss'] * 0.1
diff = abs(sc_snf['total'] - expected)
print("[4] SC+SNF 时 total 公式：")
print(f"    total={sc_snf['total']:.6f}  expected≈{expected:.6f}  diff={diff:.6f}")
ok4 = diff < 1e-3
print(f"    结果：{'✅ 通过' if ok4 else '❌ 未通过'}")
print()

print("=" * 70)
if ok1 and ok2 and ok3 and ok4:
    print("🎉 第二批修复全部通过。可以进 200 轮正式实验。")
else:
    print("⚠️ 有项目未通过，先别进 200 轮，把表格贴给 Codex 或我。")
print("=" * 70)

第二批 SNF 修复验收
[1] SNF off/on 是否改变 neg_count：
    core.neg_count = 12134772
    snf.neg_count  = 12121566
    Δ = -13206
    结果：✅ 通过

[2] SNF on 时 masked_positions > 0：
    snf.masked    = 13206
    sc_snf.masked = 13206
    结果：✅ 通过

[3] SNF-only 日志 SC=False SNF=True：
    sc_enabled=False  snf_enabled=True
    结果：✅ 通过

[4] SC+SNF 时 total 公式：
    total=8.057370  expected≈8.057370  diff=0.000000
    结果：✅ 通过

🎉 第二批修复全部通过。可以进 200 轮正式实验。


## diag2：λ_spatial=0 的四组消融诊断

### 为什么再跑一批

上一批 diag 四组（λ_spatial=0.1）：

| name | ARI | NMI | spatial | λ_sp | SC | SNF |
|---|---|---|---|---|---|---|
| core | 0.2892 | 0.3536 | 0 | 0.0 | F | F |
| sc | 0.1987 | 0.2192 | 0.797 | 0.1 | T | F |
| snf | 0.1995 | 0.2203 | 0.796 | 0.1 | F | T |
| sc_snf | 0.1988 | 0.2191 | 0.798 | 0.1 | T | T |

三组开启空间机制的版本 ARI 全部掉到 0.199 左右，几乎完全一样。
但 SC 和 SNF 是两个完全不同的机制：
- SC 改视图权重（w_std 0.0048 → 0.0103）
- SNF 改负样本 mask（neg_count 12134772 → 12121566）

作用路径不同，结果完全一致，说明掉分不是 SC/SNF 各自的问题，
而是它们共同携带的 `λ_sp=0.1` 的 spatial_loss 项。

### 关键：λ_spatial 应该 = 0

审计报告结论：
- SC 通过 MGCL 内部视图权重生效
- SNF 通过负样本 mask 生效
- 两者都不需要 `λ_spatial > 0`

我上一批加 λ_sp=0.1 是错的。这一批改回 λ_sp=0，
只让 SC/SNF 通过 MGCL 生效，才是正确的消融设计。

### 本批目的

在 λ_sp=0 下，四组 ARI：
- 都 ≈ 0.289：SC/SNF 真实无收益
- 某组 > 0.289：那组机制真实有效
- 某组 < 0.289：那组机制真实有害

无论哪种，都能得出可解释的消融表。

### 固定参数

- seed=0, epochs=50, fine_dim=32
- t=1.0, λ_mgcl=1.0, λ_cluster=0.1
- spatial_start_epoch=1（core 用 999）
- **λ_spatial=0（本批关键）**

### 与历史结果的关系

最初那次三组消融（core 0.2894 / SC 0.2883 / SC+SNF 0.2880）
其实就是在 λ_sp=0 下跑的，但当时：
- 没有 sc_enabled / snf_enabled 日志，无法确认开关是否生效
- 单次 50 轮，噪声可能大于 0.001 的差异

本批跑在修复后的代码上，有完整诊断字段，结果更可信。

新 Cell A：生成 diag2 四组配置（λ_sp 全部 = 0）

In [5]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import os, yaml
from copy import deepcopy

DATA_ROOT = "/kaggle/input/datasets/wuvdji/smgc-data"
BASE_CFG = "configs/hlna1_best.yaml"

with open(BASE_CFG) as f:
    base = yaml.safe_load(f)

os.makedirs("configs/generated", exist_ok=True)

def make(name, sc, snf, lam_spatial, start_epoch, epochs=50):
    cfg = deepcopy(base)
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = DATA_ROOT
    cfg["data"]["label_key"] = "Spatial_Label"
    cfg["output"] = {"root": "results"}
    cfg.pop("dir", None)
    cfg["training"]["epochs"] = epochs
    cfg["training"]["spatial_start_epoch"] = start_epoch
    cfg["model"]["fine_dim"] = 32
    cfg["model"]["cluster_temperature"] = 1.0
    cfg["loss"]["lambda_cluster"] = 0.1
    cfg["loss"]["lambda_mgcl"] = 1.0
    cfg["loss"]["lambda_spatial"] = lam_spatial
    cfg["spatial"] = {
        "enabled": bool(sc or snf),
        "consistency_weighting": bool(sc),
        "negative_filter": bool(snf),
    }
    path = f"configs/generated/{name}.yaml"
    with open(path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False, allow_unicode=True)
    print(f"✅ {path}  SC={sc} SNF={snf} λ_sp={lam_spatial}")

make("diag2_hlna1_core",   sc=False, snf=False, lam_spatial=0.0, start_epoch=999)
make("diag2_hlna1_sc",     sc=True,  snf=False, lam_spatial=0.0, start_epoch=1)
make("diag2_hlna1_snf",    sc=False, snf=True,  lam_spatial=0.0, start_epoch=1)
make("diag2_hlna1_sc_snf", sc=True,  snf=True,  lam_spatial=0.0, start_epoch=1)

/kaggle/working/SpaMGCL/SpaMGCL
✅ configs/generated/diag2_hlna1_core.yaml  SC=False SNF=False λ_sp=0.0
✅ configs/generated/diag2_hlna1_sc.yaml  SC=True SNF=False λ_sp=0.0
✅ configs/generated/diag2_hlna1_snf.yaml  SC=False SNF=True λ_sp=0.0
✅ configs/generated/diag2_hlna1_sc_snf.yaml  SC=True SNF=True λ_sp=0.0


新 Cell B：跑 diag2 四组

In [6]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import os, subprocess, json

NAMES = ["diag2_hlna1_core", "diag2_hlna1_sc",
         "diag2_hlna1_snf", "diag2_hlna1_sc_snf"]

for name in NAMES:
    cfg_path = f"configs/generated/{name}.yaml"
    res_path = f"results/{name}/metrics.json"
    if os.path.exists(res_path):
        print("⏭️ 已存在，跳过", name)
        continue
    print("=" * 70, "\n🚀", name)
    ret = subprocess.run(
        ["python", "experiments/run_exp.py", "--config", cfg_path],
        cwd="/kaggle/working/SpaMGCL/SpaMGCL", check=False,
    )
    print("返回码:", ret.returncode)

/kaggle/working/SpaMGCL/SpaMGCL
🚀 diag2_hlna1_core


/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.887e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400129 | rec=0.242314 | mgcl=8.157816 | cluster=2.950293 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=2.620e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232629 | mgcl=8.156867 | cluster=2.949508 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.570e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch 00

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411674 | rec=0.252383 | mgcl=8.159291 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=4.644e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400130 | rec=0.242314 | mgcl=8.157817 | cluster=2.950293 | spatial_loss=0.031559 | lambda_spatial=0 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=2.656e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232628 | mgcl=8.156868 | cluster=2.949508 | spatial_loss=0.027400 | lambda_spatial=0 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=4.687e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch 004/0

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.410588 | rec=0.252383 | mgcl=8.158205 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0 | sc_enabled=False | snf_enabled=True | mgcl_weight_std=4.887e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.399042 | rec=0.242314 | mgcl=8.156729 | cluster=2.950293 | spatial_loss=0.031563 | lambda_spatial=0 | sc_enabled=False | snf_enabled=True | mgcl_weight_std=2.621e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.388407 | rec=0.232629 | mgcl=8.155778 | cluster=2.949508 | spatial_loss=0.027400 | lambda_spatial=0 | sc_enabled=False | snf_enabled=True | mgcl_weight_std=4.570e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.962 | gradC=0.000e+00

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.410587 | rec=0.252383 | mgcl=8.158204 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0 | sc_enabled=True | snf_enabled=True | mgcl_weight_std=4.644e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.399041 | rec=0.242314 | mgcl=8.156728 | cluster=2.950293 | spatial_loss=0.031559 | lambda_spatial=0 | sc_enabled=True | snf_enabled=True | mgcl_weight_std=2.656e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.388409 | rec=0.232628 | mgcl=8.155780 | cluster=2.949508 | spatial_loss=0.027400 | lambda_spatial=0 | sc_enabled=True | snf_enabled=True | mgcl_weight_std=4.687e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.962 | gradC=0.000e+00
ep

新 Cell C：对照诊断 diag2

In [7]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import json, pandas as pd

NAMES = ["diag2_hlna1_core", "diag2_hlna1_sc",
         "diag2_hlna1_snf", "diag2_hlna1_sc_snf"]
rows = []

for name in NAMES:
    d = json.load(open(f"results/{name}/metrics.json"))
    h = d["loss_history"][-1]
    rows.append({
        "name": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "total": round(h.get("total", -1), 4),
        "mgcl": round(h.get("sample_contrastive", -1), 4),
        "w_std": round(h.get("mgcl_weight_std", -1), 5),
        "spatial": round(h.get("spatial_loss", -1), 6),
        "λ_sp": h.get("lambda_spatial", -1),
        "SC": h.get("sc_enabled"),
        "SNF": h.get("snf_enabled"),
        "neg_count": h.get("neg_count", -1),
        "masked": h.get("snf_masked_positions", -1),
    })

df = pd.DataFrame(rows)
pd.set_option("display.width", 240)
print(df.to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
              name    ARI    NMI  total   mgcl   w_std  spatial  λ_sp    SC   SNF  neg_count  masked
  diag2_hlna1_core 0.2890 0.3533 7.5940 7.2260 0.00481 0.000000   0.0 False False   12134772       0
    diag2_hlna1_sc 0.2884 0.3532 7.5869 7.2190 0.02420 1.731110   0.0  True False   12134772       0
   diag2_hlna1_snf 0.2895 0.3538 7.5915 7.2235 0.00481 1.713959   0.0 False  True   12121566   13206
diag2_hlna1_sc_snf 0.2881 0.3529 7.5831 7.2152 0.02423 1.731054   0.0  True  True   12121566   13206


## diag2 结论：HLN-A1 上 SC/SNF 是真实零效应

### 四组结果（λ_spatial=0，seed=0，50 轮）

| name | ARI | NMI | w_std | neg_count | masked | SC | SNF |
|---|---|---|---|---|---|---|---|
| core | 0.2890 | 0.3533 | 0.00481 | 12134772 | 0 | F | F |
| sc | 0.2884 | 0.3532 | **0.02420** | 12134772 | 0 | T | F |
| snf | 0.2895 | 0.3538 | 0.00481 | **12121566** | **13206** | F | T |
| sc_snf | 0.2881 | 0.3529 | 0.02423 | 12121566 | 13206 | T | T |

### 两个机制确实生效

- **SC 生效**：w_std 从 0.0048 → 0.0242，**5 倍变化**，MGCL 视图权重被显著改变
- **SNF 生效**：neg_count 少了 13206，masked_positions=13206，负样本 mask 被真正应用

### 但 ARI 全落在 0.288–0.290

四组差异最大 0.0014，**小于 seed 波动**。这不是"SC 有害"，是"SC/SNF 零效应"。

### 与之前 λ_sp=0.1 那批的区别

- λ_sp=0.1 那批：三组开启空间机制全部掉到 0.199（-0.09），**那是 λ_spatial 本身的锅**
- λ_sp=0 本批：四组全在 0.289（±0.0015），**SC/SNF 本身无影响**

### 为什么不再纠结 HLN-A1

HLN-A1 的空间图 nnz=13206，平均每 spot 只有约 3.8 个邻居，**空间结构本身较弱**。SC/SNF 想利用的空间信息有限，零效应是可解释的。

### 下一步：在 E18.5 上重复四组

- E18.5 是主结果数据集（ARI≈0.50，是 HLN-A1 的 1.7 倍）
- 空间结构更强，SC/SNF 更可能起作用
- 论文消融表至少需要两个数据集支撑
- 用完全相同配置（λ_sp=0），只改数据集和 fine_dim/t/λ_mgcl 为 E18.5 最佳

新 Cell D：生成 E18.5 diag 四组配置

In [5]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import os, yaml
from copy import deepcopy

DATA_ROOT = "/kaggle/input/datasets/wuvdji/smgc-data"
BASE_CFG = "configs/e185_best.yaml"   # 或 e185_ablation_core.yaml

with open(BASE_CFG) as f:
    base = yaml.safe_load(f)

os.makedirs("configs/generated", exist_ok=True)

def make(name, sc, snf, lam_spatial, start_epoch, epochs=50):
    cfg = deepcopy(base)
    cfg["experiment"]["name"] = name
    cfg["data"]["root"] = DATA_ROOT
    cfg["data"]["label_key"] = "Combined_Clusters"
    cfg["output"] = {"root": "results"}
    cfg.pop("dir", None)
    cfg["training"]["epochs"] = epochs
    cfg["training"]["spatial_start_epoch"] = start_epoch
    cfg["model"]["fine_dim"] = 128
    cfg["model"]["cluster_temperature"] = 0.3
    cfg["loss"]["lambda_cluster"] = 1.0
    cfg["loss"]["lambda_mgcl"] = 3.0
    cfg["loss"]["lambda_spatial"] = lam_spatial
    cfg["spatial"] = {
        "enabled": bool(sc or snf),
        "consistency_weighting": bool(sc),
        "negative_filter": bool(snf),
    }
    path = f"configs/generated/{name}.yaml"
    with open(path, "w") as f:
        yaml.dump(cfg, f, default_flow_style=False, allow_unicode=True)
    print(f"✅ {path}  SC={sc} SNF={snf} λ_sp={lam_spatial}")

make("diag2_e185_core",   sc=False, snf=False, lam_spatial=0.0, start_epoch=999)
make("diag2_e185_sc",     sc=True,  snf=False, lam_spatial=0.0, start_epoch=1)
make("diag2_e185_snf",    sc=False, snf=True,  lam_spatial=0.0, start_epoch=1)
make("diag2_e185_sc_snf", sc=True,  snf=True,  lam_spatial=0.0, start_epoch=1)

/kaggle/working/SpaMGCL/SpaMGCL
✅ configs/generated/diag2_e185_core.yaml  SC=False SNF=False λ_sp=0.0
✅ configs/generated/diag2_e185_sc.yaml  SC=True SNF=False λ_sp=0.0
✅ configs/generated/diag2_e185_snf.yaml  SC=False SNF=True λ_sp=0.0
✅ configs/generated/diag2_e185_sc_snf.yaml  SC=True SNF=True λ_sp=0.0


新 Cell E：跑 E18.5 四组

In [6]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

import os, subprocess

NAMES = ["diag2_e185_core", "diag2_e185_sc",
         "diag2_e185_snf", "diag2_e185_sc_snf"]

for name in NAMES:
    cfg_path = f"configs/generated/{name}.yaml"
    res_path = f"results/{name}/metrics.json"
    if os.path.exists(res_path):
        print("⏭️ 已存在，跳过", name); continue
    print("=" * 70, "\n🚀", name)
    subprocess.run(
        ["python", "experiments/run_exp.py", "--config", cfg_path],
        cwd="/kaggle/working/SpaMGCL/SpaMGCL", check=False,
    )

/kaggle/working/SpaMGCL/SpaMGCL
🚀 diag2_e185_core
Dataset: E18.5 | spots=2129 | device=cuda
Modalities: RNA, ATAC | label=Combined_Clusters
Spatial graph: shape=(2129, 2129), nnz=7784
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=23.166977 | rec=0.160814 | mgcl=7.668721 | cluster=3.298035 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=8.967e-03 | neg_count=4530512 | snf_masked_positions=0 | effC=13.976 | gradC=0.000e+00
epoch 002/050 | total=23.126158 | rec=0.148333 | mgcl=7.659275 | cluster=3.297572 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=7.602e-03 | neg_count=4530512 | snf_masked_positions=0 | effC=13.982 | gradC=0.000e+00
epoch 003/050 | total=23.084677 | rec=0.137188 | mgcl=7.649163 | cluster=3.297174 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=7.515e-03 | neg_count=4530512 | snf_mas

新 Cell F：对照诊断 E18.5

In [7]:
%cd /kaggle/working/SpaMGCL/SpaMGCL
import json, pandas as pd

NAMES = ["diag2_e185_core", "diag2_e185_sc",
         "diag2_e185_snf", "diag2_e185_sc_snf"]
rows = []
for name in NAMES:
    d = json.load(open(f"results/{name}/metrics.json"))
    h = d["loss_history"][-1]
    rows.append({
        "name": name,
        "ARI": round(d["ARI"], 4),
        "NMI": round(d["NMI"], 4),
        "w_std": round(h.get("mgcl_weight_std", -1), 5),
        "spatial": round(h.get("spatial_loss", -1), 6),
        "λ_sp": h.get("lambda_spatial", -1),
        "SC": h.get("sc_enabled"),
        "SNF": h.get("snf_enabled"),
        "neg_count": h.get("neg_count", -1),
        "masked": h.get("snf_masked_positions", -1),
    })
df = pd.DataFrame(rows)
pd.set_option("display.width", 240)
print(df.to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
             name    ARI    NMI   w_std  spatial  λ_sp    SC   SNF  neg_count  masked
  diag2_e185_core 0.4963 0.4719 0.00041 0.000000   0.0 False False    4530512       0
    diag2_e185_sc 0.4918 0.4692 0.02404 1.912687   0.0  True False    4530512       0
   diag2_e185_snf 0.4940 0.4724 0.00044 1.908612   0.0 False  True    4522728    7784
diag2_e185_sc_snf 0.5000 0.4728 0.02373 1.925670   0.0  True  True    4522728    7784


### 核心消融 Step 1：生成 4 组 YAML 配置

在 HLN-A1、E18.5 上各生成两组变体：
- `coreabl_hlna1_no_sample`：HLN-A1，去掉样本对比（lambda_mgcl=0）
- `coreabl_hlna1_no_cluster`：HLN-A1，去掉簇对比（lambda_cluster=0）
- `coreabl_e185_no_sample`：E18.5，去掉样本对比
- `coreabl_e185_no_cluster`：E18.5，去掉簇对比

SC/SNF 全关，λ_spatial=0，其他参数沿用各自数据集最佳配置。
只生成配置，不跑训练。

In [7]:
%cd /kaggle/working/SpaMGCL/SpaMGCL
import os, yaml
from copy import deepcopy

DATA_ROOT = "/kaggle/input/datasets/wuvdji/smgc-data"
os.makedirs("configs/core_abl", exist_ok=True)

# HLN-A1 最佳配置：dim32, t1.0, mgcl1.0, cluster0.1
# E18.5 最佳配置：dim128, t0.3, mgcl3.0, cluster1.0
DATASETS = {
    "hlna1": {
        "base": "configs/hlna1_best.yaml",
        "dataset": "HLN-A1", "label_key": "Spatial_Label",
        "n_clusters": 10, "fine_dim": 32,
        "temp": 1.0, "mgcl": 1.0, "cluster": 0.1,
    },
    "e185": {
        "base": "configs/e185_best.yaml",
        "dataset": "E18.5", "label_key": "Combined_Clusters",
        "n_clusters": 14, "fine_dim": 128,
        "temp": 0.3, "mgcl": 3.0, "cluster": 1.0,
    },
}

VARIANTS = {
    "no_sample":  {"lambda_mgcl": 0.0},
    "no_cluster": {"lambda_cluster": 0.0},
}

for ds_name, s in DATASETS.items():
    for vname, patch in VARIANTS.items():
        with open(s["base"]) as f:
            cfg = yaml.safe_load(f)
        cfg = deepcopy(cfg)
        name = f"coreabl_{ds_name}_{vname}"
        cfg["experiment"]["name"] = name
        cfg["experiment"]["dataset"] = s["dataset"]
        cfg["data"]["root"] = DATA_ROOT
        cfg["data"]["label_key"] = s["label_key"]
        cfg["output"] = {"root": "results"}
        cfg.pop("dir", None)
        cfg["training"]["epochs"] = 50
        cfg["training"]["spatial_start_epoch"] = 999
        cfg["model"]["fine_dim"] = s["fine_dim"]
        cfg["model"]["cluster_temperature"] = s["temp"]
        cfg["loss"]["lambda_mgcl"] = s["mgcl"]
        cfg["loss"]["lambda_cluster"] = s["cluster"]
        cfg["loss"]["lambda_spatial"] = 0.0
        cfg["spatial"] = {"enabled": False, "consistency_weighting": False, "negative_filter": False}
        for k, v in patch.items():
            cfg["loss"][k] = v
        path = f"configs/core_abl/{name}.yaml"
        with open(path, "w") as f:
            yaml.dump(cfg, f, default_flow_style=False, allow_unicode=True)
        print(f"✅ {path}")

/kaggle/working/SpaMGCL/SpaMGCL
✅ configs/core_abl/coreabl_hlna1_no_sample.yaml
✅ configs/core_abl/coreabl_hlna1_no_cluster.yaml
✅ configs/core_abl/coreabl_e185_no_sample.yaml
✅ configs/core_abl/coreabl_e185_no_cluster.yaml


### 核心消融 Step 2：跑 4 组训练

读取上一步生成的 YAML，依次跑 50 轮。每组约 5–10 分钟，4 组约 20–40 分钟。

跑完对比 baseline（diag2_hlna1_core / diag2_e185_core），看去掉核心损失后 ARI/NMI 掉多少。

In [8]:
%cd /kaggle/working/SpaMGCL/SpaMGCL
import os, subprocess

NAMES = [f"coreabl_{ds}_{v}" for ds in ["hlna1", "e185"] for v in ["no_sample", "no_cluster"]]

for name in NAMES:
    cfg_path = f"configs/core_abl/{name}.yaml"
    res_path = f"results/{name}/metrics.json"
    if os.path.exists(res_path):
        print("⏭️ 已存在，跳过", name); continue
    print("=" * 70, "\n🚀", name)
    subprocess.run(["python", "experiments/run_exp.py", "--config", cfg_path],
                   cwd="/kaggle/working/SpaMGCL/SpaMGCL", check=False)
    import json, pandas as pd

rows = []
for ds in ["hlna1", "e185"]:
    baseline_name = "diag2_" + ds + "_core"  # 已有结果
    try:
        d = json.load(open(f"results/{baseline_name}/metrics.json"))
        rows.append({"dataset": ds, "variant": "baseline", "ARI": round(d["ARI"], 4), "NMI": round(d["NMI"], 4)})
    except FileNotFoundError:
        print(f"⚠️ 缺 {baseline_name}，跳过")
    for v in ["no_sample", "no_cluster"]:
        name = f"coreabl_{ds}_{v}"
        try:
            d = json.load(open(f"results/{name}/metrics.json"))
            rows.append({"dataset": ds, "variant": v, "ARI": round(d["ARI"], 4), "NMI": round(d["NMI"], 4)})
        except FileNotFoundError:
            rows.append({"dataset": ds, "variant": v, "status": "未完成"})

df = pd.DataFrame(rows)
print(df.to_string(index=False))

/kaggle/working/SpaMGCL/SpaMGCL
🚀 coreabl_hlna1_no_sample


/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=0.252383 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.887e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=0.242171 | rec=0.242171 | mgcl=8.159149 | cluster=2.950280 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=5.133e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=0.232364 | rec=0.232364 | mgcl=8.159019 | cluster=2.949491 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=5.364e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch 00

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411675 | rec=0.252383 | mgcl=8.159292 | cluster=2.951158 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.887e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400130 | rec=0.242314 | mgcl=8.157817 | cluster=2.950293 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=2.620e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232629 | mgcl=8.156867 | cluster=2.949508 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.570e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch 00

## 核心组件消融结论：MGCL 双对比损失是性能的决定性来源

### 四组结果（λ_spatial=0，SC/SNF 全关，seed=0，50 轮）

| 数据集 | 配置 | ARI | NMI |
|---|---|---|---|
| HLN-A1 | baseline | 0.2890 | 0.3533 |
| HLN-A1 | w/o 样本对比 | 0.0518 | 0.0622 |
| HLN-A1 | w/o 簇对比 | 0.0492 | 0.0969 |
| E18.5 | baseline | 0.4963 | 0.4719 |
| E18.5 | w/o 样本对比 | 0.1344 | 0.2602 |
| E18.5 | w/o 簇对比 | 0.0183 | 0.0205 |

> 注：baseline 数字取自 diag2_hlna1_core / diag2_e185_core（同一配置、同一 seed、同一轮数）。

### 降幅

| 数据集 | 去掉样本对比 | 去掉簇对比 |
|---|---|---|
| HLN-A1 ARI | −82% | −83% |
| HLN-A1 NMI | −82% | −73% |
| E18.5 ARI | −73% | −96% |
| E18.5 NMI | −45% | −96% |

### 三个观察

- **去掉样本对比**：两个数据集 ARI 掉 73–82%，性能崩塌。
- **去掉簇对比**：两个数据集 ARI 掉 83–96%，E18.5 上几乎退化为随机（0.0183）。
- **两种损失缺一不可**：样本对比提供实例级判别，簇对比提供语义级结构，二者不能互相替代。

### 对论文的意义

这一组消融直接回答了审稿人最核心的两个问题：

1. **框架核心是什么？** —— MGCL 双对比损失（样本级 + 簇级）。
2. **核心组件有用吗？** —— 去掉任何一个都导致灾难性下降，是性能的决定性来源。

结合此前 SC/SNF 消融（在 HLN-A1 上零效应、在 E18.5 上微弱正收益），修正后的论文叙事为：

> 我们提出 SpaMGCL，一个基于多粒度对比学习的空间多组学聚类框架。
> **核心组件（样本级 + 簇级对比损失）经消融验证是性能的决定性来源。**
> 在此基础上，我们进一步探索了两种轻量级空间增强策略（SC 和 SNF），
> 实验表明其增量收益较小且方向不一致，揭示了空间约束在此类任务中的适用边界。

### 论文表格建议

**表 2：组件消融表**（可直接使用）

| 组件 | HLN-A1 ARI | HLN-A1 NMI | E18.5 ARI | E18.5 NMI |
|---|---|---|---|---|
| Full model | 0.2890 | 0.3533 | 0.4963 | 0.4719 |
| w/o 样本对比 | 0.0518 | 0.0622 | 0.1344 | 0.2602 |
| w/o 簇对比 | 0.0492 | 0.0969 | 0.0183 | 0.0205 |
| w/o SC | 0.2884 | 0.3532 | 0.4918 | 0.4692 |
| w/o SNF | 0.2895 | 0.3538 | 0.4940 | 0.4724 |

### 消融实验矩阵（最终版）

| 数据集 | SC/SNF 四组消融 | 核心组件消融 | 200 轮主结果 |
|---|---|---|---|
| HLN-A1 | ✅ | ✅ | 要跑 |
| E18.5 | ✅ | ✅ | 要跑 |
| HLN-D1 | ❌ 不用 | ❌ 不用 | 要跑 |
| S2-E15 | ❌ 不用 | ❌ 不用 | 要跑 |
| S2-E18 | ❌ 不用 | ❌ 不用 | 要跑 |
| Simulation | ❌ 不用 | ❌ 不用 | 暂不处理 |


## 修正后的计划：重跑 12 组 + 立刻备份
第 1 步：重跑 12 组 50 轮

In [6]:
%cd /kaggle/working/SpaMGCL/SpaMGCL

# 1) 生成 12 组配置（如果 configs 还在就跳过生成，直接跑）
import os, yaml, json, subprocess
from copy import deepcopy

DATA_ROOT = "/kaggle/input/datasets/wuvdji/smgc-data"
os.makedirs("configs/ablation_all", exist_ok=True)

DATASETS = {
    "hlna1": {
        "base": "configs/hlna1_best.yaml",
        "dataset": "HLN-A1", "label_key": "Spatial_Label",
        "n_clusters": 10, "fine_dim": 32,
        "temp": 1.0, "mgcl": 1.0, "cluster": 0.1,
    },
    "e185": {
        "base": "configs/e185_best.yaml",
        "dataset": "E18.5", "label_key": "Combined_Clusters",
        "n_clusters": 14, "fine_dim": 128,
        "temp": 0.3, "mgcl": 3.0, "cluster": 1.0,
    },
}

# 四组 SC/SNF + 两组核心消融
VARIANTS = {
    "core":         {"sc": False, "snf": False, "mgcl": None, "cluster": None},
    "sc":           {"sc": True,  "snf": False, "mgcl": None, "cluster": None},
    "snf":          {"sc": False, "snf": True,  "mgcl": None, "cluster": None},
    "sc_snf":       {"sc": True,  "snf": True,  "mgcl": None, "cluster": None},
    "no_sample":    {"sc": False, "snf": False, "mgcl": 0.0,  "cluster": None},
    "no_cluster":   {"sc": False, "snf": False, "mgcl": None, "cluster": 0.0},
}

all_names = []
for ds_name, s in DATASETS.items():
    for vname, v in VARIANTS.items():
        with open(s["base"]) as f:
            cfg = yaml.safe_load(f)
        cfg = deepcopy(cfg)
        name = f"abl_{ds_name}_{vname}"
        all_names.append(name)
        cfg["experiment"]["name"] = name
        cfg["experiment"]["dataset"] = s["dataset"]
        cfg["data"]["root"] = DATA_ROOT
        cfg["data"]["label_key"] = s["label_key"]
        cfg["output"] = {"root": "results"}
        cfg.pop("dir", None)
        cfg["training"]["epochs"] = 50
        cfg["training"]["spatial_start_epoch"] = 1 if (v["sc"] or v["snf"]) else 999
        cfg["model"]["fine_dim"] = s["fine_dim"]
        cfg["model"]["cluster_temperature"] = s["temp"]
        cfg["loss"]["lambda_mgcl"]    = s["mgcl"]    if v["mgcl"]    is None else v["mgcl"]
        cfg["loss"]["lambda_cluster"] = s["cluster"] if v["cluster"] is None else v["cluster"]
        cfg["loss"]["lambda_spatial"] = 0.0
        cfg["spatial"] = {
            "enabled": bool(v["sc"] or v["snf"]),
            "consistency_weighting": bool(v["sc"]),
            "negative_filter": bool(v["snf"]),
        }
        path = f"configs/ablation_all/{name}.yaml"
        with open(path, "w") as f:
            yaml.dump(cfg, f, default_flow_style=False, allow_unicode=True)

print(f"✅ 生成 {len(all_names)} 组配置")

# 2) 依次跑
for name in all_names:
    cfg_path = f"configs/ablation_all/{name}.yaml"
    res_path = f"results/{name}/metrics.json"
    if os.path.exists(res_path):
        print(f"⏭️ 已存在，跳过 {name}")
        continue
    print("=" * 70, f"\n🚀 {name}")
    subprocess.run(
        ["python", "experiments/run_exp.py", "--config", cfg_path],
        cwd="/kaggle/working/SpaMGCL/SpaMGCL", check=False,
    )

/kaggle/working/SpaMGCL/SpaMGCL
✅ 生成 12 组配置
🚀 abl_hlna1_core


/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411675 | rec=0.252383 | mgcl=8.159292 | cluster=2.951158 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.887e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400129 | rec=0.242314 | mgcl=8.157816 | cluster=2.950293 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=2.620e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232629 | mgcl=8.156867 | cluster=2.949508 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.570e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch 00

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411674 | rec=0.252383 | mgcl=8.159291 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=4.644e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400129 | rec=0.242314 | mgcl=8.157816 | cluster=2.950293 | spatial_loss=0.031559 | lambda_spatial=0 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=2.656e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232628 | mgcl=8.156868 | cluster=2.949508 | spatial_loss=0.027400 | lambda_spatial=0 | sc_enabled=True | snf_enabled=False | mgcl_weight_std=4.687e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch 004/0

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.410588 | rec=0.252383 | mgcl=8.158205 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0 | sc_enabled=False | snf_enabled=True | mgcl_weight_std=4.887e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.399042 | rec=0.242314 | mgcl=8.156729 | cluster=2.950293 | spatial_loss=0.031563 | lambda_spatial=0 | sc_enabled=False | snf_enabled=True | mgcl_weight_std=2.621e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.388408 | rec=0.232629 | mgcl=8.155779 | cluster=2.949508 | spatial_loss=0.027400 | lambda_spatial=0 | sc_enabled=False | snf_enabled=True | mgcl_weight_std=4.570e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.962 | gradC=0.000e+00

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.410587 | rec=0.252383 | mgcl=8.158204 | cluster=2.951158 | spatial_loss=0.037660 | lambda_spatial=0 | sc_enabled=True | snf_enabled=True | mgcl_weight_std=4.644e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.399041 | rec=0.242314 | mgcl=8.156728 | cluster=2.950293 | spatial_loss=0.031559 | lambda_spatial=0 | sc_enabled=True | snf_enabled=True | mgcl_weight_std=2.656e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.388409 | rec=0.232628 | mgcl=8.155780 | cluster=2.949508 | spatial_loss=0.027400 | lambda_spatial=0 | sc_enabled=True | snf_enabled=True | mgcl_weight_std=4.687e-03 | neg_count=12121566 | snf_masked_positions=13206 | effC=9.962 | gradC=0.000e+00
ep

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=0.252383 | rec=0.252383 | mgcl=8.159292 | cluster=2.951158 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.887e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=0.242171 | rec=0.242171 | mgcl=8.159149 | cluster=2.950280 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=5.133e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=0.232364 | rec=0.232364 | mgcl=8.159019 | cluster=2.949491 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=5.364e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch 00

/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}
/kaggle/working/SpaMGCL/SpaMGCL/src/data/dataset.py:191: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  adatas = {modality: anndata.read_h5ad(path) for modality, path in paths.items()}


Dataset: HLN-A1 | spots=3484 | device=cuda
Modalities: RNA, ADT | label=Spatial_Label
Spatial graph: shape=(3484, 3484), nnz=13206
Learning rates: backbone=0.001 | cluster_head=0.008 (8x)
epoch 001/050 | total=8.411676 | rec=0.252383 | mgcl=8.159293 | cluster=2.951158 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.887e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.950 | gradC=0.000e+00
epoch 002/050 | total=8.400129 | rec=0.242314 | mgcl=8.157816 | cluster=2.950293 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=2.620e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.956 | gradC=0.000e+00
epoch 003/050 | total=8.389496 | rec=0.232629 | mgcl=8.156867 | cluster=2.949508 | spatial_loss=0.000000 | lambda_spatial=0 | sc_enabled=False | snf_enabled=False | mgcl_weight_std=4.570e-03 | neg_count=12134772 | snf_masked_positions=0 | effC=9.962 | gradC=0.000e+00
epoch 00

跑完后立刻下载，用这个 cell 打 zip

In [7]:
%cd /kaggle/working/SpaMGCL/SpaMGCL
!zip -r /kaggle/working/ablation_results.zip results/ configs/ablation_all/
print("打包完成：/kaggle/working/ablation_results.zip")

/kaggle/working/SpaMGCL/SpaMGCL
  adding: results/ (stored 0%)
  adding: results/abl_hlna1_sc_snf/ (stored 0%)
  adding: results/abl_hlna1_sc_snf/config.yaml (deflated 50%)
  adding: results/abl_hlna1_sc_snf/metrics.json (deflated 84%)
  adding: results/abl_e185_sc/ (stored 0%)
  adding: results/abl_e185_sc/config.yaml (deflated 52%)
  adding: results/abl_e185_sc/metrics.json (deflated 84%)
  adding: results/abl_hlna1_core/ (stored 0%)
  adding: results/abl_hlna1_core/config.yaml (deflated 50%)
  adding: results/abl_hlna1_core/metrics.json (deflated 85%)
  adding: results/p1_hlna1/ (stored 0%)
  adding: results/p1_hlna1/config.yaml (deflated 46%)
  adding: results/p1_hlna1/metrics.json (deflated 81%)
  adding: results/abl_e185_no_cluster/ (stored 0%)
  adding: results/abl_e185_no_cluster/config.yaml (deflated 52%)
  adding: results/abl_e185_no_cluster/metrics.json (deflated 86%)
  adding: results/abl_hlna1_no_sample/ (stored 0%)
  adding: results/abl_hlna1_no_sample/config.yaml (deflat

跑完 + 下载后，本地解压提取 concat_z

In [9]:
import json, pandas as pd
from pathlib import Path

results_dir = Path("results")  # 解压后的目录

rows = []
for name in sorted(results_dir.iterdir()):
    p = name / "metrics.json"
    if not p.exists():
        continue
    with open(p) as f:
        d = json.load(f)

    q_ari = d.get("ARI")
    q_nmi = d.get("NMI")

    rep = d.get("representation_metrics", {})
    cz = rep.get("concat_z", {})
    concat_ari = cz.get("ARI")
    concat_nmi = cz.get("NMI")

    rows.append({
        "name": name.name,
        "Q_argmax_ARI": round(q_ari, 4) if q_ari else None,
        "Q_argmax_NMI": round(q_nmi, 4) if q_nmi else None,
        "concat_z_ARI": round(concat_ari, 4) if concat_ari else None,
        "concat_z_NMI": round(concat_nmi, 4) if concat_nmi else None,
    })

df = pd.DataFrame(rows)
pd.set_option("display.width", 240)
print(df.to_string(index=False))
df.to_csv("ablation_concat_z_summary.csv", index=False)

                name  Q_argmax_ARI  Q_argmax_NMI  concat_z_ARI  concat_z_NMI
       abl_e185_core        0.4961        0.4717        0.3959        0.5115
 abl_e185_no_cluster        0.0194        0.0218        0.4446        0.5381
  abl_e185_no_sample        0.1344        0.2602        0.2018        0.2889
         abl_e185_sc        0.4915        0.4694        0.3943        0.5060
     abl_e185_sc_snf        0.4994        0.4738        0.3769        0.4967
        abl_e185_snf        0.4931        0.4721        0.3669        0.4969
      abl_hlna1_core        0.2892        0.3536        0.2403        0.3355
abl_hlna1_no_cluster        0.0492        0.0969        0.2414        0.3365
 abl_hlna1_no_sample        0.0518        0.0622        0.0633        0.1216
        abl_hlna1_sc        0.2881        0.3529        0.2417        0.3374
    abl_hlna1_sc_snf        0.2881        0.3531        0.2414        0.3373
       abl_hlna1_snf        0.2895        0.3538        0.2386        0.3346

## 核心组件消融：concat_z 视角的修正解读

### 之前解读的错误

上一版 markdown 写“去掉簇对比导致表征崩塌”，这是错的。
Q_argmax 崩塌不等于表征崩塌——Q 是聚类头输出，去掉簇对比后
gradC=0，聚类头没有训练信号，Q 的输出自然退化为垃圾。

要判断表征好坏，必须看 concat_z。

### 修正后的结果

| 数据集 | 配置 | Q_argmax ARI | concat_z ARI |
|---|---|---|---|
| E18.5 | core | 0.4961 | 0.3959 |
| E18.5 | no_cluster | 0.0194 | 0.4446 |
| E18.5 | no_sample | 0.1344 | 0.2018 |
| HLN-A1 | core | 0.2892 | 0.2403 |
| HLN-A1 | no_cluster | 0.0492 | 0.2414 |
| HLN-A1 | no_sample | 0.0518 | 0.0633 |

### 两个损失的分工

- **样本对比损失**：决定表征质量。去掉后 concat_z 掉 49–74%。
- **簇对比损失**：不直接改善表征（去掉后 concat_z 持平或略好），
  但是端到端聚类头 Q 的唯一训练信号（去掉后 Q 崩塌）。

### 论文叙事

> 样本级对比损失是表征学习的决定性来源；
> 簇级对比损失不直接改善表征，但为端到端聚类头提供训练信号。
> 两者分工互补，共同决定端到端聚类性能。